tilings/algorithms/obstruction_inferral.py

```diff
@@ -0,0 +1,149 @@
+import abc
+from itertools import chain, product
+
+from comb_spec_searcher import InferralRule
+from permuta import Perm, PermSet
+from tilings import Obstruction
+
+
+class ObstructionInferral(abc.ABC):
+    """
+    Algorithm to compute the tiling created by adding all obstructions from
+    `self.potential_new_obs()` that can be added to the tiling.
+    """
+    def __init__(self, tiling):
+        self._tiling = tiling
+
+    @abc.abstractmethod
+    def potential_new_obs(self):
+        """
+        Return an iterable of new obstruction that should be added to the
+        tiling if possible
+        """
+        pass
+
+    def new_obs(self):
+        """
+        Returns the list of new obstruction that can be added to the tiling.
+        """
+        if hasattr(self, '_new_obs'):
+            return self._new_obs
+        newobs = []
+        merged_tiling = self._tiling.merge()
+        for ob in sorted(self.potential_new_obs(), key=len):
+            if self.can_add_obstruction(ob, merged_tiling):
+                newobs.append(ob)
+                merged_tiling = merged_tiling.__class__(
+                    obstructions=chain(merged_tiling.obstructions, (ob,)),
+                    requirements=merged_tiling.requirements,
+                    remove_empty=False)
```

Comment:
Rather than creating a new tiling, it is probably faster to just check if it avoids all of the newobs. In fact, you probably want to do between line 33 and 34.

Reply:
Did it